cpu/stm32f4/periph/eth.c

```diff
@@ -0,0 +1,540 @@
+/*
+ * Copyright (C) 2016 TriaGnoSys GmbH
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for more
+ * details.
+ */
+
+/**
+ * @ingroup     cpu_stm32f4
+ * @{
+ *
+ * @file
+ * @brief       Low-level ETH driver implementation
+ *
+ * @author      Víctor Ariño <victor.arino@triagnosys.com>
+ *
+ * @}
+ */
+
+#include "cpu.h"
+#include "mutex.h"
+#include "cib.h"
+#include "periph_conf.h"
+#include "periph/gpio.h"
+
+#include "board.h"
+
+#include "net/netdev2.h"
+#include "net/netdev2/eth.h"
+#include "net/eui64.h"
+#include "net/ethernet.h"
+#include "net/netstats.h"
+
+#include "eth-phy.h"
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+#define SDEBUG(...) DEBUG("eth: " __VA_ARGS__)
+
+#include <string.h>
+
+#if ETH_NUMOF
+
+/* Set the value of the divider with the clock configured */
+#if !defined(CLOCK_CORECLOCK) || CLOCK_CORECLOCK < (20000000U)
+#error This peripheral requires a CORECLOCK of at least 20MHz
+#elif CLOCK_CORECLOCK < (35000000U)
+#define CLOCK_RANGE ETH_MACMIIAR_CR_Div16
+#elif CLOCK_CORECLOCK < (60000000U)
+#define CLOCK_RANGE ETH_MACMIIAR_CR_Div26
+#elif CLOCK_CORECLOCK < (100000000U)
+#define CLOCK_RANGE ETH_MACMIIAR_CR_Div42
+#elif CLOCK_CORECLOCK < (150000000U)
+#define CLOCK_RANGE ETH_MACMIIAR_CR_Div62
+#else /* CLOCK_CORECLOCK < (20000000U) */
+#define CLOCK_RANGE ETH_MACMIIAR_CR_Div102
+#endif /* CLOCK_CORECLOCK < (20000000U) */
+
+#define MIN(a, b) ((a < b) ? a : b)
+
+/* Internal flags for the DMA descriptors */
+#define DESC_OWN           (0x80000000)
+#define RX_DESC_FL         (0x3FFF0000)
+#define RX_DESC_FS         (0x00000200)
+#define RX_DESC_LS         (0x00000100)
+#define RX_DESC_RCH        (0x00004000)
+#define TX_DESC_TCH        (0x00100000)
+#define TX_DESC_IC         (0x40000000)
+#define TX_DESC_CIC        (0x00C00000)
+#define TX_DESC_LS         (0x20000000)
+#define TX_DESC_FS         (0x10000000)
+
+struct eth_dma_desc {
+    uint32_t status;
+    uint32_t control;
+    char *buffer_addr;
+    struct eth_dma_desc *desc_next;
+    uint32_t reserved1_ext;
+    uint32_t reserved2;
+    uint32_t ts_low;
+    uint32_t ts_high;
+} __attribute__((packed));
+
+typedef struct eth_dma_desc edma_desc_t;
+
+/** Update pointer to next descriptor */
+#define _next_desc(x) x = (x->desc_next)
```

Comment:
Would you mind removing this macro? We're trying to not use macros when possible.